File: solutions/Medium/Combination_Sum_0039.js
/**
 * Function: combinationSum
 * Description: Given an array of distinct integers `candidates` and a target integer `target`,
 * this function returns all unique combinations of numbers that sum to `target`. 
 * Each number in `candidates` may be used an unlimited number of times.
 *
 * @param {number[]} candidates - An array of distinct positive integers.
 * @param {number} target - The target sum to achieve.
 * @returns {number[][]} - A list of all unique combinations summing to `target`.
 */
const combinationSum = function(candidates, target) {
    let result = [];

    /**
     * Backtracking function to find valid combinations.
     * @param {number} start - The current index in the `candidates` array.
     * @param {number[]} combination - The current combination of numbers.
     * @param {number} sum - The current sum of elements in `combination`.
     */
    const backtrack = (start, combination, sum) => {
        // If sum equals target, add the combination to the result
        if (sum === target) {
            result.push([...combination]);
            return;
        }

        // If sum exceeds target, stop exploring further
        if (sum > target) return;

        // Iterate over the candidates starting from `start` index
        for (let i = start; i < candidates.length; i++) {
            combination.push(candidates[i]); // Include the current candidate
            backtrack(i, combination, sum + candidates[i]); // Recur with the same index (unlimited reuse)
            combination.pop(); // Backtrack and remove the last added element
        }
    };

    // Start backtracking with an empty combination
    backtrack(0, [], 0);

    return result;
};

// Example Test Cases
if (require.main === module) {
    console.log("Test 1: Basic case");
    console.log(combinationSum([2, 3, 6, 7], 7));
    // Expected output: [[2,2,3],[7]]

    console.log("Test 2: Multiple valid combinations");
    console.log(combinationSum([2, 3, 5], 8));
    // Expected output: [[2,2,2,2],[2,3,3],[3,5]]

    console.log("Test 3: No valid combinations");
    console.log(combinationSum([2], 1));
    // Expected output: []

    console.log("Test 4: Single candidate can form target multiple ways");
    console.log(combinationSum([3], 9));
    // Expected output: [[3,3,3]]

    console.log("Test 5: Large target with different numbers");
    console.log(combinationSum([2, 5, 10], 20));
    // Expected output: [[2,2,2,2,2,2,2,2,2,2],[2,2,2,2,2,2,2,2,5],[2,2,2,2,2,2,5,5],[2,2,2,2,5,5,5],[2,2,2,10,10],[2,2,5,5,5,5],[2,5,5,5,5],[5,5,5,5],[10,10]]

    console.log("Test 6: Only one candidate and it's exactly target");
    console.log(combinationSum([7], 7));
    // Expected output: [[7]]
}

// Complexity analysis
// Time Complexity: O(2^n) - Exponential due to generating all possible subsets.
// Space Complexity: O(2^n) - Due to storing all valid subsets in the result array.

export default combinationSum;

File: solutions/Medium/Combination_Sum_0039.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        """
        Function: combinationSum
        Description: Finds all unique combinations of candidates where chosen numbers sum to target.

        Parameters:
        - candidates (List[int]): A list of distinct integers.
        - target (int): The sum target value.

        Returns:
        - List[List[int]]: A list of all unique combinations that sum to target.
        """
        result = []

        def backtrack(start, combination, current_sum):
            # Base case: if current sum reaches the target, add to result
            if current_sum == target:
                result.append(list(combination))
                return
            if current_sum > target:
                return  # Stop the recursion if sum exceeds target

            for i in range(start, len(candidates)):
                # Include candidates[i] and recursively explore
                combination.append(candidates[i])
                # Allow same element reuse
                backtrack(i, combination, current_sum + candidates[i])
                combination.pop()  # Backtrack

        backtrack(0, [], 0)

        return result


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Simple case with multiple valid combinations
    print("Test Case 1: candidates = [2,3,6,7], target = 7")
    print("Expected Output: [[2,2,3],[7]] | Actual Output:",
          solution.combinationSum([2, 3, 6, 7], 7))

    # Test case 2: Multiple combinations with a different set
    print("Test Case 2: candidates = [2,3,5], target = 8")
    print("Expected Output: [[2,2,2,2],[2,3,3],[3,5]] | Actual Output:",
          solution.combinationSum([2, 3, 5], 8))

    # Test case 3: No valid combination
    print("Test Case 3: candidates = [2], target = 1")
    print("Expected Output: [] | Actual Output:",
          solution.combinationSum([2], 1))

    # Test case 4: Large number in candidates
    print("Test Case 4: candidates = [2,7,9], target = 14")
    print("Expected Output: [[2,2,2,2,2,2,2],[7,7]] | Actual Output:",
          solution.combinationSum([2, 7, 9], 14))

    # Test case 5: Single element repeated multiple times
    print("Test Case 5: candidates = [5], target = 10")
    print("Expected Output: [[5,5]] | Actual Output:",
          solution.combinationSum([5], 10))

    # Test case 6: Edge case with large elements
    print("Test Case 6: candidates = [3,5,8], target = 11")
    print("Expected Output: [[3,3,5],[3,8]] | Actual Output:",
          solution.combinationSum([3, 5, 8], 11))

"""
Time Complexity:
- O(2^N) → The worst case explores all subsets due to repeated choices.

Space Complexity:
- O(target/min(candidates)) → Depth of the recursion tree.
"""
